cmd/perflock/perflock_test.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+const (
+	// The sleep duration must be short enough that the tests don't become a chore
+	// to run, but long enough that we can avoid the noise of too much antagonist
+	// (OS, other procs, ...) jitter.
+	sleepDuration = 500 * time.Millisecond
+)
+
+// Integration-style tests. These tests emulate what a command-line user would
+// do. E.g.: running a daemon and a number of clients. To do so while minimizing
+// the hassle needed to run test (`go test .` is sufficient), the test setup
+// uses subprocesses of itself to function as daemon, client and user-supplied
+// programs.
+//
+// The environment variables affecting this behaviour are GO_TEST_MODE and
+// GO_TEST_PROGRAM_MODE. See the comments below.
+//
+// About OS support: perflock effectively only works fully on Linux, as it uses
+// OS-specific interfaces (e.g. for the CPU governor). The tests use abstract
+// names (starting with @) for the UNIX domain socket to listen on. We don't
+// bother skipping for non-Linux, as that will hopefully make it clear what
+// should be fixed to those who are interested.
+func TestMain(m *testing.M) {
+	switch os.Getenv("GO_TEST_MODE") {
+	case "": // Run tests (top-level).
+		os.Exit(m.Run())
+
+	case "perflock": // Act like a perflock.
+		// If GO_TEST_PROGRAM_MODE is set, we're a perflock client, and main() will
+		// be spawn os.Args[0] again to run in program mode. To activate program
+		// mode on the next execution, set GO_TEST_MODE=program.
+		if os.Getenv("GO_TEST_PROGRAM_MODE") != "" {
+			os.Setenv("GO_TEST_MODE", "program")
+		}
+		// Remove the flags registered by the testing package.
+		//
+		// NOTE: flag.Parse() has not been called yet (we're in TestMain), according
+		// to https://pkg.go.dev/testing
+		flag.CommandLine = flag.NewFlagSet("dummy", flag.ExitOnError) // This call is required because otherwise flags panics, if args are set between flag.Parse calls
+		main()
+
+	case "program": // Act like a "normal" program.
+		switch pmode := os.Getenv("GO_TEST_PROGRAM_MODE"); pmode {
+		case "sleeper":
+			sleeper()
+		default:
+			log.Fatalf("unknown program mode %q", pmode)
+		}
+	}
+}
+
+func waitForDaemon(t *testing.T, ctx context.Context, socket string) bool {
+	for {
+		c, err := net.Dial("unix", socket)
+		if err != nil {
+			// TODO(aktau): Deal with errors that aren't not found?
+			select {
+			case <-ctx.Done():
+				return false
+			case <-time.After(100 * time.Millisecond):
+				continue
+			}
+		}
+		c.Close()
+		return true
+	}
+}
+
+func TestExclusive(t *testing.T) {
+	t.Parallel()
+
+	socket := fmt.Sprintf("@perflock.%d.%s", os.Getpid(), funcname())
```

Comment:
I'm not sure you really need the abstract socket. Couldn't this just be a regular UNIX domain socket in a directory created by t.TempDir()? That would also avoid the need for the PID and funcname()

Reply:
At this point I've:

 1. Split the support for abstract domain sockets into a separate commit.
 2. Use them in tests on Linux. This serves as a way of testing this support as well. Tests should also work on macOS, as on non-Linux OSen it uses socket files.